File: react-native/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/RNPlacementProviderManager.kt
package com.storylyplacementreactnative.common

import android.content.Context
import android.util.Log
import com.appsamurai.storyly.core.data.model.STRDataPayload
import com.appsamurai.storyly.core.data.model.STRDataSource
import com.appsamurai.storyly.core.data.model.product.STRProductInformation
import com.appsamurai.storyly.core.listener.provider.STRProviderListener
import com.appsamurai.storyly.core.listener.provider.STRProviderProductListener
import com.appsamurai.storyly.placement.data.provider.PlacementDataProvider


object RNPlacementProviderManager {

    private val providers = mutableMapOf<String, RNPlacementProviderWrapper>()
    private val lock = Any()

    fun createProvider(context: Context, id: String): RNPlacementProviderWrapper {
        synchronized(lock) {
            Log.d("[RNPlacementProviderManager]", "Create provider: $id")
            val wrapper = RNPlacementProviderWrapper(context, id)
            providers[id] = wrapper
            return wrapper
        }
    }

    fun getProvider(id: String): RNPlacementProviderWrapper? {
        synchronized(lock) {
            return providers[id]
        }
    }

    fun destroyProvider(id: String) {
        synchronized(lock) {
            Log.d("[RNPlacementProviderManager]", "Destroy provider: $id")
            providers.remove(id)
        }
    }
}

class RNPlacementProviderWrapper(
    private val context: Context,
    val id: String
) {

    var provider: PlacementDataProvider = PlacementDataProvider(context)

    internal var sendEvent: ((String, RNPlacementProviderEventType, String) -> Unit)? = null

    fun configure(configJson: String) {
        val parsedConfig = RNPlacementDataConverter.decodeFromJson(configJson)
        if (parsedConfig == null) {
            Log.e("[RNPlacementProviderWrapper]", "Failed to parse config JSON")
            return
        }

        setupProvider(parsedConfig)
    }

    private fun setupProvider(config: Map<String, Any?>) {
        val token = config["token"] as? String ?: return

        Log.d("[RNPlacementProviderWrapper]", "Configuring provider with token: $token")

        val placementConfig = RNPlacementDataConverter.createSTRPlacementConfig(config, token)

        provider.apply {
          listener = object : STRProviderListener {
            override fun onLoad(data: STRDataPayload, dataSource: STRDataSource) {
              val eventData = mapOf(
                "dataSource" to dataSource.value,
              )
              val eventJson = RNPlacementDataConverter.encodeToJson(eventData)
              Log.d("[RNPlacementProviderWrapper]", "STRProviderListener:onLoad: $eventJson")
              sendEvent?.invoke(id, RNPlacementProviderEventType.ON_LOAD, eventJson ?: "")
            }

            override fun onLoadFail(errorMessage: String) {
              val eventData = mapOf(
                "errorMessage" to errorMessage,
              )
              val eventJson = RNPlacementDataConverter.encodeToJson(eventData)
              Log.d("[RNPlacementProviderWrapper]", "STRProviderListener:onLoadFail: $eventJson")
              sendEvent?.invoke(id, RNPlacementProviderEventType.ON_LOAD_FAIL, eventJson ?: "")
            }
          }
          productListener = object : STRProviderProductListener {
            override fun onHydration(products: List<STRProductInformation>) {
              val productsMap = products.map { RNPlacementDataConverter.createSTRProductInformationMap(it) }
              val eventData = mapOf(
                "products" to productsMap,
              )
              val eventJson = RNPlacementDataConverter.encodeToJson(eventData)
              Log.d("[RNPlacementProviderWrapper]", "STRProviderProductListener:onHydration: $eventJson")
              sendEvent?.invoke(id, RNPlacementProviderEventType.ON_HYDRATION, eventJson ?: "")
            }
          }

          this.config = placementConfig
        }
    }

    fun hydrateProducts(productsJson: String) {
        Log.d("[RNPlacementProviderWrapper]", "hydrateProducts: $productsJson")
        val products = RNPlacementDataConverter.parseProductItems(productsJson)
        provider.hydrateProducts(products)
    }

    fun hydrateWishlist(productsJson: String) {
        Log.d("[RNPlacementProviderWrapper]", "hydrateWishlist: $productsJson")
        val products = RNPlacementDataConverter.parseProductItems(productsJson)
        provider.hydrateWishlist(products)
    }

    fun updateCart(cartJson: String) {
        Log.d("[RNPlacementProviderWrapper]", "updateCart: $cartJson")
        val cart = RNPlacementDataConverter.parseCart(cartJson)
        if (cart != null) {
            provider.updateCart(cart)
        }
    }
}

File: react-native/storyly-placement-react-native/ios/Common/RNPlacementProviderManager.swift
import Foundation
import StorylyPlacement
import StorylyCore

@objc public class RNPlacementProviderManager: NSObject {
    
    @objc public static let shared = RNPlacementProviderManager()
    
    private var providers: [String: RNPlacementProviderWrapper] = [:]
    private let lock = NSLock()
    
    private override init() {
        super.init()
    }
    
    @objc public func createProvider(id: String) -> RNPlacementProviderWrapper {
        lock.lock()
        defer { lock.unlock() }
        
        print("[RNPlacementProviderManager] Create provider: \(id)")
        let wrapper = RNPlacementProviderWrapper(id: id)
        providers[id] = wrapper
        return wrapper
    }
    
    @objc public func getProvider(id: String) -> RNPlacementProviderWrapper? {
        lock.lock()
        defer { lock.unlock() }
        
        return providers[id]
    }
    
    @objc public func destroyProvider(id: String) {
        lock.lock()
        defer { lock.unlock() }
        
        print("[RNPlacementProviderManager] Destroy provider: \(id)")
        providers.removeValue(forKey: id)
    }
}

@objc public class RNPlacementProviderWrapper: NSObject {
    
    @objc public let id: String
    @objc public lazy var provider: PlacementDataProvider = {
        return PlacementDataProvider()
    }()
    
    @objc public var sendEvent: ((String, RNPlacementProviderEventType, String) -> Void)?
    
    init(id: String) {
        self.id = id
        super.init()
    }
    
    @objc public func configure(configJson: String) {
        guard let parsedConfig = decodeFromJson(configJson) else {
            print("[RNPlacementProviderWrapper] Failed to parse config JSON")
            return
        }
        
        setupProvider(config: parsedConfig)
    }
    
    private func setupProvider(config: [String: Any]) {
        guard let token = config["token"] as? String else {
            print("[RNPlacementProviderWrapper] Token not found in config")
            return
        }
        
        print("[RNPlacementProviderWrapper] Configuring provider with token: \(token)")
        
        let placementConfig = decodeSTRPlacementConfig(config, token: token)
        
        provider.delegate = STRProviderListenerImpl(wrapper: self)
        provider.productDelegate = STRProviderProductListenerImpl(wrapper: self)
        provider.config = placementConfig
    }
    
    @objc public func hydrateProducts(productsJson: String) {
        guard let dict = decodeFromJson(productsJson),
              let productsArray = dict["products"] as? [[String: Any]] else {
            return
        }
        
        print("[RNPlacementProviderWrapper] hydrateProducts: \(productsJson)")
        
        let products = productsArray.compactMap { decodeSTRProductItem($0) }
        provider.hydrateProducts(products: products)
    }
    
    @objc public func hydrateWishlist(productsJson: String) {
        guard let dict = decodeFromJson(productsJson),
              let productsArray = dict["products"] as? [[String: Any]] else {
            return
        }
        
        print("[RNPlacementProviderWrapper] hydrateWishlist: \(productsJson)")
        
        let products = productsArray.compactMap { decodeSTRProductItem($0) }
        provider.hydrateWishlist(products: products)
    }
    
    @objc public func updateCart(cartJson: String) {
        guard let dict = decodeFromJson(cartJson),
              let cartDict = dict["cart"] as? [String: Any],
              let cart = decodeSTRCart(cartDict) else {
            return
        }
        
        print("[RNPlacementProviderWrapper] updateCart: \(cartJson)")
        
        provider.updateCart(cart: cart)
    }
}

// MARK: - STRProviderListener Implementation

private class STRProviderListenerImpl: NSObject, STRProviderDelegate {
    weak var wrapper: RNPlacementProviderWrapper?
    
    init(wrapper: RNPlacementProviderWrapper) {
        self.wrapper = wrapper
    }
    
    func onLoad(data: STRDataPayload, dataSource: STRDataSource) {
        guard let wrapper = wrapper else { return }
        
        let eventData: [String: Any] = [
            "data": encodeDataPayload(data),
            "dataSource": dataSource.description
        ]
        
        if let eventJson = encodeToJson(eventData) {
            print("[RNPlacementProviderWrapper] STRProviderListener:onLoad: \(eventJson)")
            wrapper.sendEvent?(wrapper.id, .onLoad, eventJson)
        }
    }
    
    func onLoadFail(errorMessage: String) {
        guard let wrapper = wrapper else { return }
        
        let eventData: [String: Any] = [
            "errorMessage": errorMessage
        ]
        
        if let eventJson = encodeToJson(eventData) {
            print("[RNPlacementProviderWrapper] STRProviderListener:onLoadFail: \(eventJson)")
          wrapper.sendEvent?(wrapper.id, .onLoadFail, eventJson)
        }
    }
}

// MARK: - STRProviderProductListener Implementation

private class STRProviderProductListenerImpl: NSObject, STRProviderProductDelegate {
    weak var wrapper: RNPlacementProviderWrapper?
    
    init(wrapper: RNPlacementProviderWrapper) {
        self.wrapper = wrapper
    }
    
    func onHydration(products: [STRProductInformation]) {
        guard let wrapper = wrapper else { return }
        
        let eventData: [String: Any] = [
            "products": products.map { encodeSTRProductInformation($0) }
        ]
        
        if let eventJson = encodeToJson(eventData) {
            print("[RNPlacementProviderWrapper] STRProviderProductListener:onHydration: \(eventJson)")
            wrapper.sendEvent?(wrapper.id, .onHydration, eventJson)
        }
    }
}

